src/gc-arena/src/context.rs

```diff
@@ -196,6 +221,21 @@ impl Context {
                         // the main list and destruct it, otherwise it should be black, and we
                         // simply turn it white again.
                         if sweep.flags.color() == GcColor::White {
+                            if sweep.flags.has_weak_ref() {
+                                // Keep the `GcBox` as part of the linked list if we traced a weak pointer
+                                // to it. The weak pointer still needs access to the `GcBox` to be able to
+                                // check if the object is still alive. We can only deallocate the `GcBox`,
+                                // once there are no weak pointers left.
+                                self.sweep_prev.set(Some(sweep_ptr));
+                                sweep.flags.set_has_weak_ref(false);
+                                if sweep.flags.alive() {
+                                    sweep.flags.set_alive(false);
+                                    // SAFETY: Since this object is white, that means there are no more strong pointers
+                                    // to this object, only weak pointers, so we can safely drop its contents.
+                                    core::ptr::drop_in_place(sweep.value.get());
+                                }
+                                continue;
```

Comment:
nit: I don't like the `continue` because this method is enormous and I have trouble knowing where it goes.. the *ideal* solution is to break all of the phases into their own methods but you don't have to do that, for now can the next block just be in an `} else {` so that it doesn't look *different* from the other logic? It's not any different but it looks special because of the `continue`. I know the indentation is insane right now..

Reply:
I've refactored this to use an `else` block.